SG14_test/inplace_function_test.cpp

```diff
@@ -429,6 +429,77 @@ static void RvalueRefParameter()
     g(std::make_unique<int>(42));
 }
 
+struct NoDefaultCtor
+{
+    int val;
+    explicit NoDefaultCtor(int v) : val{v} {}
+};
+
+int overloaded_function(stdext::inplace_function<int()>&& exec)
+{
+    return exec();
+}
+
+int overloaded_function(stdext::inplace_function<int(int)>&& exec)
+{
+    return exec(int{42});
+}
+
+using C_Int1 = int();
+using C_Int2 = int(int);
+using C_Void = void(int&);
+
+using stdext::inplace_function_detail::is_invocable_r;
+
+static_assert(is_invocable_r<int, C_Int1>::value, "IV A");
+static_assert(! is_invocable_r<int, C_Int2>::value, "IV B");
+static_assert(! is_invocable_r<int, C_Void>::value, "IV C");
```

Comment:
I'd like to see additional checks somewhere in here for

    static_assert(is_invocable_r<void, C_Int1>::value);
    static_assert(is_invocable_r<long, C_Int1>::value);
    static_assert(is_invocable_r<char, C_Int1>::value);

(testing widening and narrowing conversions, and the "conversion" to `void`). Also,

    static_assert(! is_invocable_r<int, C_Void, int&>::value);

(testing the conversion *from* `void` to `int`, which should definitely not be allowed).

Finally, I think we should have regression tests for both `is_invocable` and `is_convertible`:

    static_assert(is_invocable_r<const int&, int()>::value);
    static_assert(is_convertible<int(*)(), inplace_function<const int&()>>::value);

[These tests do currently pass for `std::function`](https://godbolt.org/z/OP23Ae) but my understanding is that at least the second line is expected to fail after the adoption of whatever paper is resolving [LWG 2813](https://cplusplus.github.io/LWG/issue2813).

Reply:
`static_assert(is_convertible<int(*)(), inplace_function<const int&()>>::value, "E2")` fails, and I don't quite understand why.

Reply:
The culprit is the `std::is_convertible<C, inplace_function>::value` here:

    template<
        typename T,
        typename C = std::decay_t<T>,
        typename = std::enable_if_t<
            !(std::is_same<C, inplace_function>::value
            || std::is_convertible<C, inplace_function>::value)
            && inplace_function_detail::is_invocable_r<R, C, Args...>::value
        >
    >
    inplace_function(T&& closure)

This is already problematic in master:

    static_assert(std::is_convertible<int(*)(), stdext::inplace_function<int()>>::value, "");

The problem is that you're using `is_convertible<C, ipf>` as a SFINAE condition in order to decide whether to _make_ `C` convertible to `ipf`. I think this is only working by accident. Anyway, in this specific case, `is_convertible` needs to decide whether there's any implicit constructor `inplace_function(int(*arg)())` (it knows there's no implicit conversion operator from `int(*)()` to `inplace_function`). So it does template substitution into this SFINAE condition, which causes the instantiation of `std::is_convertible<C, inplace_function> [with C=int(*)()]`, which makes a cycle in the instantiation graph. I don't know if this even has defined behavior; but in practice what it does is make `std::is_convertible<C, inplace_function>` come out to `false_type`. And then that `false_type` gets memoized by the compiler and re-used for the condition of the `static_assert` later on. Does that make sense?

I think the solution is going to have to be to turn the conversion operators back into constructors. Which we already want to do because of #125.

Reply:
@Voultapher wrote:
> It feels like this pr is spiraling out of scope. Given that the remaining problems exist on master and are somewhat unrelated to the changes in this pr, I'd suggest merging this as is and addressing #125 in a separate pr.

I agree that this PR is spiraling out of scope. However, I strongly recommend that we fix the known issues with master (i.e. fix `std::is_convertible<int(*)(), stdext::inplace_function<int()>>`) *before* introducing new and compounded issues (i.e. enabling people to SFINAE on that already-wrong expression).

That means we should fix `std::is_convertible<int(*)(), stdext::inplace_function<int()>>` in master, by whatever means necessary (I strongly suggest "by means of fixing #125"), and delay merging this PR until it can be merged onto a solid foundation.

Reply:
Sounds reasonable, I'll get on it.

Reply:
Resolved by #157